src/Compilers/Test/Core/Metadata/ILValidation.cs

```diff
@@ -320,106 +325,122 @@
-            {
-                doc.Load(reader);
-            }
-
             var result = new Dictionary<int, string>();
 
-            if (source == null)
-            {
-                static void Add(Dictionary<int, string> dict, int key, string value)
-                    => dict[key] = dict.TryGetValue(key, out var found) ? found + value : value;
+            void add(int key, string value)
+                => result[key] = result.TryGetValue(key, out var existing) ? existing + value : value;
 
-                foreach (XmlNode entry in doc.GetElementsByTagName("sequencePoints"))
+            foreach (var e in methodXml.Descendants())
+            {
+                if (e.Name == "entry" && e.Parent.Name == "sequencePoints")
                 {
-                    foreach (XmlElement item in entry.ChildNodes)
-                    {
-                        Add(result,
-                            Convert.ToInt32(item.GetAttribute("offset"), 16),
-                            (item.GetAttribute("hidden") == "true") ? "~" : "-");
-                    }
+                    add(Convert.ToInt32(e.Attribute("offset").Value, 16), (e.Attribute("hidden")?.Value == "true") ? "~" : "-");
                 }
+            }
 
-                foreach (XmlNode entry in doc.GetElementsByTagName("asyncInfo"))
+            foreach (var e in methodXml.Descendants())
+            {
+                if (e.Name == "await" && e.Parent.Name == "asyncInfo")
                 {
-                    foreach (XmlElement item in entry.ChildNodes)
-                    {
-                        if (item.Name == "await")
-                        {
-                            Add(result, Convert.ToInt32(item.GetAttribute("yield"), 16), "<");
-                            Add(result, Convert.ToInt32(item.GetAttribute("resume"), 16), ">");
-                        }
-                        else if (item.Name == "catchHandler")
-                        {
-                            Add(result, Convert.ToInt32(item.GetAttribute("offset"), 16), "$");
-                        }
-                    }
+                    add(Convert.ToInt32(e.Attribute("yield").Value, 16), "<");
+                    add(Convert.ToInt32(e.Attribute("resume").Value, 16), ">");
+                }
+                else if (e.Name == "catchHandler" && e.Parent.Name == "asyncInfo")
+                {
+                    add(Convert.ToInt32(e.Attribute("offset").Value, 16), "$");
                 }
             }
-            else
-            {
-                static void AddTextual(Dictionary<int, string> dict, int key, string value)
-                    => dict[key] = dict.TryGetValue(key, out var found) ? found + ", " + value : "// " + value;
 
-                foreach (XmlNode entry in doc.GetElementsByTagName("asyncInfo"))
+            return result;
+        }
+
+        public static Dictionary<int, string> GetSequencePointMarkers(XElement methodXml, Func<string, SourceText> getSource)
```

Comment:
Feels like these two `GetSequencePointMarkers` methods are 90% identical. Can we factor them into a single implementation?

Reply:
They are actually quite different. I don't see benefit of a single implementation, especially since the `GetSequencePointMarkers` is only there for existing tests and should ideally be removed at some point.

Reply:
> They are actually quite different

I'm not convinced. I did a diff between the two and the only two differences are the ordering of two `if` blocks and which strings we decide to print out (either fixed strings like "~" or strings with some source).
I'll let second reviewer chime in.

Reply:
Is there a reason why the two `foreach` blocks are swapped?

> the `GetSequencePointMarkers` is only there for existing tests and should ideally be removed at some point

I'm not sure we will ever refactor all the tests to get rid of this method, so it makes sense to me to share the code if possible.

Reply:
> Is there a reason why the two foreach blocks are swapped?

I don't know, the logic was implemented in two separate code paths before my change. I just moved it to a separate method. I really don't think there is any value in trying to unify them.

Reply:
> I'm not sure we will ever refactor all the tests to get rid of this method, so it makes sense to me to share the code if possible.

There is not that many of these tests.